File: src/nightlight.ts
import WinReg from 'winreg'

const STATE_KEY_PATH =
  '\\Software\\Microsoft\\Windows\\CurrentVersion\\CloudStore\\Store\\DefaultAccount\\Current\\default$windows.data.bluelightreduction.bluelightreductionstate\\windows.data.bluelightreduction.bluelightreductionstate'

const SETTINGS_KEY_PATH =
  '\\Software\\Microsoft\\Windows\\CurrentVersion\\CloudStore\\Store\\DefaultAccount\\Current\\default$windows.data.bluelightreduction.settings\\windows.data.bluelightreduction.settings'

const MIN_KELVIN = 1200 // Maximum warmth (100% strength)
const MAX_KELVIN = 6500 // Neutral (0% strength)

/**
 * A class for inspecting Windows 10/11's Night Light feature.
 */
export class NightLight {
  private _registryKey: WinReg.Registry
  private _settingsKey: WinReg.Registry

  constructor() {
    this._registryKey = new WinReg({
      hive: WinReg.HKCU,
      key: STATE_KEY_PATH,
    })
    this._settingsKey = new WinReg({
      hive: WinReg.HKCU,
      key: SETTINGS_KEY_PATH,
    })
  }

  supported(): boolean {
    return this._registryKey != null && this._settingsKey != null
  }

  private getData(): Promise<WinReg.RegistryItem> {
    return new Promise<WinReg.RegistryItem>((resolve, reject) =>
      this._registryKey.get('Data', (err, result) =>
        err ? reject(err) : resolve(result)
      )
    )
  }

  private getSettingsData(): Promise<WinReg.RegistryItem> {
    return new Promise<WinReg.RegistryItem>((resolve, reject) =>
      this._settingsKey.get('Data', (err, result) =>
        err ? reject(err) : resolve(result)
      )
    )
  }

  async enabled(): Promise<boolean> {
    if (!this.supported()) return false
    const data = await this.getData()
    if (!data) return false
    const bytes = hexToBytes(data.value)
    return bytes[18] === 0x15 // 21 in decimal
  }

  async enable(): Promise<void> {
    if (this.supported() && !(await this.enabled())) this.toggle()
  }

  async disable(): Promise<void> {
    if (this.supported() && (await this.enabled())) this.toggle()
  }

  async toggle(): Promise<void> {
    let newData: number[]
    const enabled = await this.enabled()
    const rawData = await this.getData()
    const data = hexToBytes(rawData.value)

    if (enabled) {
      newData = new Array(41).fill(0)
      newData.splice(0, 22, ...data.slice(0, 22))
      newData.splice(23, 43 - 25, ...data.slice(25, 43))
      newData[18] = 0x13
    } else {
      newData = new Array(43).fill(0)
      newData.splice(0, 22, ...data.slice(0, 22))
      newData.splice(25, 41 - 23, ...data.slice(23, 41))
      newData[18] = 0x15
      newData[23] = 0x10
      newData[24] = 0x00
    }

    for (let i = 10; i < 15; i++) {
      if (newData[i] !== 0xff) {
        newData[i]++
        break
      }
    }

    const newDataHex = bytesToHex(newData)
    return new Promise<void>((resolve, reject) =>
      this._registryKey.set('Data', WinReg.REG_BINARY, newDataHex, err => {
        err ? reject(err) : resolve()
      })
    )
  }

  async getStrength(): Promise<number> {
    if (!this.supported()) return 0
    const data = await this.getSettingsData()
    if (!data) return 0
    const bytes = hexToBytes(data.value)
    const kelvin = this.bytesToKelvin(bytes[0x23], bytes[0x24])
    return this.kelvinToPercentage(kelvin)
  }

  async setStrength(percentage: number): Promise<void> {
    if (!this.supported()) return
    
    // Clamp percentage between 0-100
    percentage = Math.max(0, Math.min(100, percentage))
    
    // Convert percentage to kelvin
    const kelvin = this.percentageToKelvin(percentage)
    
    const rawData = await this.getSettingsData()
    if (!rawData) return
    const data = hexToBytes(rawData.value)

    // Calculate bytes using the PowerShell script's formula
    const tempHi = Math.floor(kelvin / 64)
    const tempLo = ((kelvin - (tempHi * 64)) * 2) + 128

    // Update strength bytes (indices 0x23, 0x24)
    data[0x23] = tempLo
    data[0x24] = tempHi

    // Update timestamp bytes
    for (let i = 10; i < 15; i++) {
      if (data[i] !== 0xff) {
        data[i]++
        break
      }
    }

    const newDataHex = bytesToHex(data)
    return new Promise<void>((resolve, reject) =>
      this._settingsKey.set('Data', WinReg.REG_BINARY, newDataHex, err => {
        err ? reject(err) : resolve()
      })
    )
  }

  private bytesToKelvin(loTemp: number, hiTemp: number): number {
    // Convert bytes back to kelvin using the inverse of the PowerShell formula
    return (hiTemp * 64) + ((loTemp - 128) / 2)
  }

  private kelvinToPercentage(kelvin: number): number {
    // Inverse linear mapping from kelvin to percentage
    return 100 - ((kelvin - MIN_KELVIN) / (MAX_KELVIN - MIN_KELVIN)) * 100
  }

  private percentageToKelvin(percentage: number): number {
    // Linear mapping from percentage to kelvin
    return MAX_KELVIN - (percentage / 100) * (MAX_KELVIN - MIN_KELVIN)
  }
}

// Convert a hex string to a byte array
function hexToBytes(hex: string): number[] {
  let bytes = []
  for (let c = 0; c < hex.length; c += 2)
    bytes.push(parseInt(hex.substr(c, 2), 16))
  return bytes
}

// Convert a byte array to a hex string
function bytesToHex(bytes: number[]): string {
  let hex = []
  for (let i = 0; i < bytes.length; i++) {
    let current = bytes[i] < 0 ? bytes[i] + 256 : bytes[i]
    hex.push((current >>> 4).toString(16))
    hex.push((current & 0xf).toString(16))
  }
  return hex.join('')
}

// Example usage:
// const nightLight = new NightLight()
// console.log('Supported:', nightLight.supported())
// console.log('Enabled:', await nightLight.enabled())
// console.log('Toggling')
// await nightLight.toggle()
// console.log('Enabled:', await nightLight.enabled())

File: src/nightlight.py
"""
NightLight Python Module
Control Windows 10/11's Night Light feature programmatically.
"""

import winreg
from typing import Optional

# Registry paths for Night Light settings
STATE_KEY_PATH = r"Software\Microsoft\Windows\CurrentVersion\CloudStore\Store\DefaultAccount\Current\default$windows.data.bluelightreduction.bluelightreductionstate\windows.data.bluelightreduction.bluelightreductionstate"
SETTINGS_KEY_PATH = r"Software\Microsoft\Windows\CurrentVersion\CloudStore\Store\DefaultAccount\Current\default$windows.data.bluelightreduction.settings\windows.data.bluelightreduction.settings"

# Temperature constants
MIN_KELVIN = 1200  # Maximum warmth (100% strength)
MAX_KELVIN = 6500  # Neutral (0% strength)


class NightLight:
    """A class for controlling Windows 10/11's Night Light feature."""

    def __init__(self):
        """Initialize the NightLight controller."""
        self._state_key_path = STATE_KEY_PATH
        self._settings_key_path = SETTINGS_KEY_PATH

    def supported(self) -> bool:
        """Check if Night Light feature is supported on this system."""
        try:
            winreg.OpenKey(winreg.HKEY_CURRENT_USER, self._state_key_path, 0, winreg.KEY_READ)
            winreg.OpenKey(winreg.HKEY_CURRENT_USER, self._settings_key_path, 0, winreg.KEY_READ)
            return True
        except FileNotFoundError:
            return False

    def _get_data(self) -> Optional[bytes]:
        """Get the registry data for Night Light state."""
        try:
            key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, self._state_key_path, 0, winreg.KEY_READ)
            data, _ = winreg.QueryValueEx(key, "Data")
            winreg.CloseKey(key)
            return data
        except Exception:
            return None

    def _get_settings_data(self) -> Optional[bytes]:
        """Get the registry data for Night Light settings."""
        try:
            key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, self._settings_key_path, 0, winreg.KEY_READ)
            data, _ = winreg.QueryValueEx(key, "Data")
            winreg.CloseKey(key)
            return data
        except Exception:
            return None

    def enabled(self) -> bool:
        """Check if Night Light is currently enabled."""
        if not self.supported():
            return False

        data = self._get_data()
        if data is None:
            return False

        return data[18] == 0x15  # 21 in decimal

    def enable(self) -> None:
        """Enable Night Light."""
        if self.supported() and not self.enabled():
            self.toggle()

    def disable(self) -> None:
        """Disable Night Light."""
        if self.supported() and self.enabled():
            self.toggle()

    def toggle(self) -> None:
        """Toggle Night Light on/off."""
        if not self.supported():
            raise RuntimeError("Night Light feature is not supported on this system.")

        enabled = self.enabled()
        data = self._get_data()

        if data is None:
            raise RuntimeError("Could not retrieve Night Light data.")

        data = bytearray(data)

        if enabled:
            # Create a 41-element array filled with zeros
            new_data = bytearray(41)

            # Copy data[0-21] to new_data[0-21]
            new_data[0:22] = data[0:22]

            # Copy data[25-42] to new_data[23-40]
            if len(data) > 25:
                copy_length = min(len(data) - 25, 43 - 25)
                new_data[23:23 + copy_length] = data[25:25 + copy_length]

            new_data[18] = 0x13
        else:
            # Create a 43-element array filled with zeros
            new_data = bytearray(43)

            # Copy data[0-21] to new_data[0-21]
            new_data[0:22] = data[0:22]

            # Copy data[23-40] to new_data[25-42]
            if len(data) > 23:
                copy_length = min(len(data) - 23, 41 - 23)
                new_data[25:25 + copy_length] = data[23:23 + copy_length]

            new_data[18] = 0x15
            new_data[23] = 0x10
            new_data[24] = 0x00

        # Increment the first byte in the range 10-14 that isn't 0xff
        for i in range(10, 15):
            if new_data[i] != 0xff:
                new_data[i] = (new_data[i] + 1) % 256
                break

        # Update the registry
        try:
            key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, self._state_key_path, 0, winreg.KEY_SET_VALUE)
            winreg.SetValueEx(key, "Data", 0, winreg.REG_BINARY, bytes(new_data))
            winreg.CloseKey(key)
        except Exception as e:
            raise RuntimeError(f"Failed to update Night Light registry data: {e}")

    def _bytes_to_kelvin(self, lo_temp: int, hi_temp: int) -> float:
        """Convert bytes back to kelvin temperature."""
        return (hi_temp * 64) + ((lo_temp - 128) / 2)

    def _kelvin_to_percentage(self, kelvin: float) -> float:
        """Convert kelvin temperature to percentage."""
        return 100 - ((kelvin - MIN_KELVIN) / (MAX_KELVIN - MIN_KELVIN)) * 100

    def _percentage_to_kelvin(self, percentage: float) -> float:
        """Convert percentage to kelvin temperature."""
        return MAX_KELVIN - (percentage / 100) * (MAX_KELVIN - MIN_KELVIN)

    def get_strength(self) -> int:
        """Get current Night Light strength as percentage (0-100)."""
        if not self.supported():
            return 0

        data = self._get_settings_data()
        if data is None:
            return 0

        # Get temperature bytes from indices 0x23 and 0x24
        temp_lo = data[0x23]
        temp_hi = data[0x24]

        # Convert bytes to kelvin
        kelvin = self._bytes_to_kelvin(temp_lo, temp_hi)

        # Convert kelvin to percentage
        return round(self._kelvin_to_percentage(kelvin))

    def set_strength(self, percentage: int) -> None:
        """
        Set Night Light strength percentage.

        Args:
            percentage: Strength value between 0 (coolest) and 100 (warmest)
        """
        if not self.supported():
            raise RuntimeError("Night Light feature is not supported on this system.")

        # Clamp percentage between 0-100
        percentage = max(0, min(100, percentage))

        # Convert percentage to kelvin
        kelvin = self._percentage_to_kelvin(percentage)

        data = self._get_settings_data()
        if data is None:
            raise RuntimeError("Could not retrieve Night Light settings data.")

        # Create a copy of the current data
        new_data = bytearray(data)

        # Calculate bytes for the temperature
        temp_hi = int(kelvin // 64)
        temp_lo = int(((kelvin - (temp_hi * 64)) * 2) + 128)

        # Update temperature bytes (indices 0x23, 0x24)
        new_data[0x23] = temp_lo
        new_data[0x24] = temp_hi

        # Update timestamp bytes
        for i in range(10, 15):
            if new_data[i] != 0xff:
                new_data[i] = (new_data[i] + 1) % 256
                break

        # Update the registry
        try:
            key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, self._settings_key_path, 0, winreg.KEY_SET_VALUE)
            winreg.SetValueEx(key, "Data", 0, winreg.REG_BINARY, bytes(new_data))
            winreg.CloseKey(key)
        except Exception as e:
            raise RuntimeError(f"Failed to update Night Light strength: {e}")


# Example usage
if __name__ == "__main__":
    night_light = NightLight()

    print(f"Supported: {night_light.supported()}")
    print(f"Enabled: {night_light.enabled()}")
    print(f"Current strength: {night_light.get_strength()}%")
